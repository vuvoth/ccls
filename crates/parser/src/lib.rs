pub mod event;
pub mod grammar;
pub mod parser;
pub mod token_kind;

pub use logos::Lexer;

pub mod input;
pub mod output;
