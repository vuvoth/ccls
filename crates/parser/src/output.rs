use crate::{event::Event, token_kind::TokenKind};

#[derive(Debug)]
pub enum Child {
    Token(usize), // position of token,
    Error(String),
    Tree(Tree),
}

#[derive(Debug)]
pub struct Tree {
    kind: TokenKind,
    children: Vec<Child>,
}

pub type Output = Tree;

impl Output {
    fn empty() -> Self {
        Tree {
            kind: TokenKind::ParserError,
            children: Vec::new(),
        }
    }

    pub fn kind(&self) -> TokenKind {
        self.kind
    }

    pub fn children(&self) -> &Vec<Child> {
        &self.children
    }
}

impl From<Vec<Event>> for Output {
    fn from(events: Vec<Event>) -> Self {
        let mut stack = Vec::new();
        if let Some((last, elements)) = events.split_last() {
            if !matches!(*last, Event::Close) {
                return Output::empty();
            }
            for event in elements {
                match event {
                    Event::Open { kind } => {
                        stack.push(Tree {
                            kind: *kind,
                            children: Vec::new(),
                        });
                    }
                    Event::Close => {
                        let tree = stack.pop().unwrap();

                        stack.last_mut().unwrap().children.push(Child::Tree(tree));
                    }
                    Event::TokenPosition(token) => {
                        stack
                            .last_mut()
                            .unwrap()
                            .children
                            .push(Child::Token(*token));
                    }
                    Event::ErrorReport(error) => {
                        stack
                            .last_mut()
                            .unwrap()
                            .children
                            .push(Child::Error(error.clone()));
                    }
                }
            }
        }
        // TODO: Make it more safe
        stack.pop().unwrap()
    }
}
