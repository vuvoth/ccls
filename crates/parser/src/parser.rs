use std::cell::Cell;

use crate::{
    event::Event, grammar::entry::Scope, input::Input, output::Output, token_kind::TokenKind,
};

pub struct Context {
    pub r_curly_count: i32,
}

pub struct Parser<'a> {
    pub(crate) input: &'a Input<'a>,
    pub context: Context,
    pos: usize,
    fuel: Cell<u32>,
    pub(crate) events: Vec<Event>,
}

#[derive(Clone, Copy, Debug)]
pub enum Marker {
    Open(usize),
    Close(usize),
}

#[derive(Debug)]
pub enum ParserError {
    InvalidEvents,
}

impl<'a> Parser<'a> {
    pub fn wrap_trivial_tokens(&mut self) -> TokenKind {
        loop {
            let kind = self.input.kind_of(self.pos);

            if kind.is_trivial() == false {
                return kind;
            }

            self.events.push(Event::Open { kind });

            self.fuel.set(256);
            self.events.push(Event::TokenPosition(self.pos));
            self.skip();

            self.events.push(Event::Close);
        }
    }

    pub fn open(&mut self) -> Marker {
        if self.events.len() > 0 {
            self.wrap_trivial_tokens();
        }

        let marker = Marker::Open(self.events.len());
        self.events.push(Event::Open {
            kind: TokenKind::Error,
        });
        marker
    }

    pub fn open_before(&mut self, marker_closed: Marker) -> Marker {
        match marker_closed {
            Marker::Close(index) => {
                let marker_opened = Marker::Open(index);
                self.events.insert(
                    index,
                    Event::Open {
                        kind: TokenKind::EOF,
                    },
                );
                marker_opened
            }
            _ => unreachable!(),
        }
    }

    pub fn close(&mut self, open_marker: Marker, kind: TokenKind) -> Marker {
        match open_marker {
            Marker::Open(index) => {
                self.events[index] = Event::Open { kind };
                self.events.push(Event::Close);
                Marker::Close(index)
            }
            _ => unreachable!(),
        }
    }

    pub fn advance(&mut self) {
        // assert!(!self.eof());
        self.fuel.set(256);
        let token = Event::TokenPosition(self.pos);
        self.events.push(token);
        self.skip();
    }

    pub fn advance_with_token(&mut self, index: usize) {
        // assert!(token.kind != TokenKind::EOF);
        if self.input.kind_of(index) != TokenKind::EOF {
            self.fuel.set(256);
            let token = Event::TokenPosition(index);
            self.events.push(token);
        }
    }

    pub fn advance_with_error(&mut self, _error: &str) {
        let m = self.open();
        // TODO: Error reporting.
        if !self.eof() {
            self.advance();
        }
        self.close(m, TokenKind::Error);
    }

    pub fn error_report(&mut self, error: String) {
        let m = self.open();

        let token = Event::ErrorReport(error);
        self.events.push(token);

        self.close(m, TokenKind::Error);
    }
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a Input) -> Self {
        Self {
            input,
            pos: 0,
            context: Context { r_curly_count: 0 },
            fuel: Cell::new(256),
            events: Vec::new(),
        }
    }

    pub fn inc_rcurly(&mut self) {
        self.context.r_curly_count += 1;
    }

    pub fn dec_rcurly(&mut self) {
        self.context.r_curly_count -= 1;
    }

    pub fn current(&mut self) -> TokenKind {
        self.wrap_trivial_tokens()
    }

    pub fn next(&mut self) -> TokenKind {
        if self.fuel.get() == 0 {
            panic!("parser is stuck");
        }
        self.fuel.set(self.fuel.get() - 1);
        if self.pos < self.input.size() {
            self.pos += 1;
            return self.input.kind_of(self.pos);
        }

        TokenKind::EOF
    }

    pub fn at(&mut self, kind: TokenKind) -> bool {
        self.current() == kind
    }

    pub fn at_any(&mut self, kinds: &[TokenKind]) -> bool {
        let current_kind = self.current();
        kinds.contains(&current_kind)
    }

    pub fn at_assign_token(&mut self) -> bool {
        let current_kind = self.current();
        current_kind.is_assign_token()
    }

    pub fn at_inline_assign_signal(&mut self) -> bool {
        let current_kind = self.current();
        current_kind.is_inline_assign_signal()
    }

    pub fn at_var_assign(&mut self) -> bool {
        let current_kind = self.current();
        current_kind.is_var_assign()
    }

    pub fn skip(&mut self) {
        self.next();
    }

    pub fn skip_if(&mut self, kinds: &[TokenKind]) {
        if self.at_any(kinds) {
            self.skip();
        }
    }

    pub fn eat(&mut self, kind: TokenKind) -> bool {
        if self.at(kind) {
            self.advance();
            return true;
        }

        false
    }

    pub fn expect_any(&mut self, kinds: &[TokenKind]) {
        let kind = self.current();
        if kinds.contains(&kind) {
            self.advance();
        } else {
            let error = format!("expect {:?} but got {:?}", kinds, kind);
            self.error_report(error);
        }
    }

    pub fn expect(&mut self, kind: TokenKind) {
        if self.at(kind) {
            self.advance();
        } else {
            let error = format!("expect {:?} but got {:?}", kind, self.current());
            self.error_report(error);
        }
    }

    pub fn eof(&mut self) -> bool {
        self.current() == TokenKind::EOF
    }
}

impl Parser<'_> {
    pub fn parsing_with_scope(input: &Input, scope: Scope) -> Output {
        let mut p = Parser::new(input);
        scope.parse(&mut p);
        Output::from(p.events)
    }

    pub fn parsing(input: &Input) -> Output {
        let c = Scope::CircomProgram;
        Parser::parsing_with_scope(input, c)
    }
}
