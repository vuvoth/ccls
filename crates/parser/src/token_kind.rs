use logos::Logos;
use serde::Serialize;

#[derive(Logos, Debug, PartialEq, Clone, Copy, Eq, PartialOrd, Ord, Hash, Serialize)]
#[allow(non_camel_case_types)]
#[repr(u16)]
pub enum TokenKind {
    // Error
    #[error]
    Error = 0,
    // Comments
    #[regex(r"//[^\r\n]*")]
    CommentLine,
    #[token("/*")]
    CommentBlockOpen,
    #[token("*/")]
    CommentBlockClose,
    // Trivial
    #[regex("[ \t]+")]
    WhiteSpace,
    #[regex(r"\r?\n")]
    EndLine,
    // Pragma
    Pragma,
    #[token("pragma")]
    PragmaKw,
    #[token("circom")]
    Circom,
    #[regex("2.[0-9].[0-9]")]
    Version,
    // Literals
    #[regex("[0-9]+")]
    Number,
    #[regex("[$_]*[a-zA-Z][a-zA-Z0-9_$]*")]
    Identifier,
    #[regex(r#""[^"]*""#)]
    CircomString,
    // Brackets
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LCurly,
    #[token("}")]
    RCurly,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,
    // Punctuation
    #[token(";")]
    Semicolon,
    #[token(",")]
    Comma,
    #[token(".")]
    Dot,
    // Boolean operators
    #[token("&&")]
    BoolAnd,
    #[token("||")]
    BoolOr,
    #[token("!")]
    Not,
    // Relational operators
    #[token("==")]
    Equal,
    #[token("!=")]
    NotEqual,
    #[token("<")]
    LessThan,
    #[token(">")]
    GreaterThan,
    #[token("<=")]
    LessThanAndEqual,
    #[token(">=")]
    GreaterThanAndEqual,
    // Arithmetic operators
    #[token("+")]
    Add,
    #[token("-")]
    Sub,
    #[token("*")]
    Mul,
    #[token("**")]
    Power,
    #[token("/")]
    Div,
    #[token("\\")]
    IntDiv,
    #[token("%")]
    Mod,
    // Combined arithmetic assignment
    #[token("+=")]
    AddAssign,
    #[token("-=")]
    SubAssign,
    #[token("*=")]
    MulAssign,
    #[token("**=")]
    PowerAssign,
    #[token("/=")]
    DivAssign,
    #[token(r"\=")]
    IntDivAssign,
    #[token("%=")]
    ModAssign,
    #[token("++")]
    UnitInc,
    #[token("--")]
    UnitDec,
    // Bitwise operators
    #[token("&")]
    BitAnd,
    #[token("|")]
    BitOr,
    #[token("~")]
    BitNot,
    #[token("^")]
    BitXor,
    #[token(">>")]
    ShiftR,
    #[token("<<")]
    ShiftL,
    // Combined bitwise assignments
    #[token("&=")]
    BitAndAssign,
    #[token("|=")]
    BitOrAssign,
    #[token("~=")]
    BitNotAssign,
    #[token("^=")]
    BitXorAssign,
    #[token(">>=")]
    ShiftRAssign,
    #[token("<<=")]
    ShiftLAssign,
    // Assign
    #[token("=")]
    Assign,
    #[token("===")]
    EqualSignal,
    #[token("-->")]
    LAssignSignal,
    #[token("==>")]
    LAssignContraintSignal,
    #[token("<--")]
    RAssignSignal,
    #[token("<==")]
    RAssignConstraintSignal,
    // Conditional expressions
    #[token("?")]
    MarkQuestion,
    #[token(":")]
    Colon,
    // Keywords
    #[token("template")]
    TemplateKw,
    #[token("function")]
    FunctionKw,
    #[token("component")]
    ComponentKw,
    #[token("main")]
    MainKw,
    #[token("public")]
    PublicKw,
    #[token("signal")]
    SignalKw,
    #[token("var")]
    VarKw,
    #[token("include")]
    IncludeKw,
    #[token("input")]
    InputKw,
    #[token("output")]
    OutputKw,
    #[token("log")]
    LogKw,
    // Statement keywords
    #[token("if")]
    IfKw,
    #[token("else")]
    ElseKw,
    #[token("for")]
    ForKw,
    #[token("while")]
    WhileKw,
    #[token("return")]
    ReturnKw,
    #[token("assert")]
    AssertKw,
    // Statements
    IfStatement,
    AssertStatement,
    LogStatement,
    ReturnStatement,
    AssignStatement,
    ForLoop,
    WhileLoop,
    // Program
    CircomProgram,
    // Function
    FunctionDef,
    FunctionName,
    // Template
    TemplateDef,
    TemplateName,
    // Signal
    SignalDecl,
    InputSignalDecl,
    OutputSignalDecl,
    SignalHeader,
    SignalIdentifier,
    // Variable
    VarDecl,
    VarIdentifier,
    // Component
    ComponentDecl,
    ComponentCall,
    ComponentIdentifier,
    SignalOfComponent,
    // Expression
    ExpressionAtom,
    Expression,
    // Complex token kind
    Block,
    ParameterList,
    Call,
    TenaryConditional,
    Condition,
    Statement,
    StatementList,
    ArrayQuery,
    ParserError,
    BlockComment,
    EOF,
    ROOT,
    __LAST,
}

impl From<u16> for TokenKind {
    #[inline]
    fn from(d: u16) -> TokenKind {
        assert!(d <= (TokenKind::__LAST as u16));
        unsafe { std::mem::transmute::<u16, TokenKind>(d) }
    }
}

impl From<rowan::SyntaxKind> for TokenKind {
    fn from(value: rowan::SyntaxKind) -> Self {
        match value {
            rowan::SyntaxKind(id) => TokenKind::from(id),
        }
    }
}

impl From<TokenKind> for u16 {
    #[inline]
    fn from(k: TokenKind) -> u16 {
        k as u16
    }
}

impl From<TokenKind> for rowan::SyntaxKind {
    fn from(kind: TokenKind) -> Self {
        Self(kind as u16)
    }
}

impl TokenKind {
    // a + 10 --> a and 10 are literals
    pub fn is_literal(self) -> bool {
        matches!(self, Self::Number | Self::Identifier)
    }

    // these tokens have the lowest priority
    // <identifier1> infix_operator <identifier2>
    // eg: a + b --> + is an infix token
    pub fn infix(self) -> Option<(u16, u16)> {
        match self {
            // arithmetic operators
            Self::Power => Some((99, 100)),
            Self::Mul | Self::Div | Self::IntDiv | Self::Mod => Some((94, 95)),
            Self::Add | Self::Sub => Some((89, 90)),
            // shift bitwise operators
            Self::ShiftL | Self::ShiftR => Some((84, 85)),
            // relational operators
            Self::LessThan
            | Self::GreaterThan
            | Self::LessThanAndEqual
            | Self::GreaterThanAndEqual => Some((79, 80)),
            Self::Equal | Self::NotEqual => Some((74, 75)),
            // other bitwise operators
            Self::BitAnd => Some((69, 70)),
            Self::BitXor => Some((64, 65)), // exclusive or
            Self::BitOr => Some((59, 60)),
            // boolean operators
            Self::BoolAnd => Some((54, 55)),
            Self::BoolOr => Some((49, 50)),
            // ----------
            // TODO: how about conditional operation ( ? : )
            // associativity: right to left [ a ? b : c --> ??? ]

            // ----------
            // associativity: right to left [ a = b = c --> a = (b = c) ]
            // DO NOT CONSIDER ASSIGMENT OPERATORS AS INFIX TOKENS
            /*
            // assignment operators
            Self::Assign
            // signal assigment operators
            | Self::EqualSignal
            | Self::LAssignSignal
            | Self::LAssignContraintSignal
            | Self::RAssignSignal
            | Self::RAssignConstraintSignal
            // bitwise asignment operators
            | Self::BitOrAssign
            | Self::BitXorAssign
            | Self::BitAndAssign
            | Self::ShiftLAssign
            | Self::ShiftRAssign
            // arithmetic asignament operators
            | Self::AddAssign
            | Self::SubAssign
            | Self::MulAssign
            | Self::DivAssign
            | Self::IntDivAssign
            | Self::ModAssign
            | Self::PowerAssign => Some((44, 45)),
            */
            // TODO: how about comma (expression separator)
            Self::Comma => Some((39, 40)),
            // not an infix operator
            _ => None,
        }
    }

    // priority: post > pre > in
    // associativity: right to left [ --!a --> --(!a) ]
    // prefix_operator <literal>
    // eg: -10, !a, ++a, --a
    pub fn prefix(self) -> Option<u16> {
        match self {
            Self::UnitDec | Self::UnitInc | Self::Sub | Self::Add | Self::Not | Self::BitNot => {
                Some(200)
            }

            _ => None,
        }
    }

    // these tokens have the highest priority
    // <literal> postfix_operator
    // eg: a[10], b++, c.att1
    pub fn postfix(self) -> Option<u16> {
        match self {
            Self::LParen // function call
            | Self::LBracket // array subscript
            | Self::Dot // attribute access
            | Self::UnitDec | Self::UnitInc => Some(300),

            _ => None,
        }
    }

    pub fn is_declaration_kw(self) -> bool {
        matches!(self, Self::VarKw | Self::ComponentKw | Self::SignalKw)
    }

    pub fn is_assign_token(self) -> bool {
        matches!(
            self,
            Self::Assign
            // signal assigment operators
            | Self::EqualSignal
            | Self::LAssignSignal
            | Self::LAssignContraintSignal
            | Self::RAssignSignal
            | Self::RAssignConstraintSignal
            // bitwise asignment operators
            | Self::BitOrAssign
            | Self::BitXorAssign
            | Self::BitAndAssign
            | Self::ShiftLAssign
            | Self::ShiftRAssign
            // arithmetic asignament operators
            | Self::AddAssign
            | Self::SubAssign
            | Self::MulAssign
            | Self::DivAssign
            | Self::IntDivAssign
            | Self::ModAssign
            | Self::PowerAssign // unit inc/dec
                                // | Self::UnitInc
                                // | Self::UnitDec
        )
    }

    pub fn is_inline_assign_signal(self) -> bool {
        matches!(
            self,
            Self::Assign | Self::RAssignSignal | Self::RAssignConstraintSignal
        )
    }

    pub fn is_var_assign(self) -> bool {
        matches!(self, Self::Assign)
    }

    pub fn is_trivial(self) -> bool {
        matches!(
            self,
            Self::WhiteSpace | Self::EndLine | Self::CommentLine | Self::BlockComment | Self::Error
        )
    }
}
