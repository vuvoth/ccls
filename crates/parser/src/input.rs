use std::ops::Range;

use logos::Lexer;

use serde::Serialize;

use crate::token_kind::TokenKind;

#[derive(Debug, PartialEq, Serialize)]
pub struct Input<'a> {
    kind: Vec<TokenKind>,
    source: &'a str,
    position: Vec<Range<usize>>,
}

impl<'a> Input<'a> {
    pub fn new(source: &'a str) -> Self {
        let mut input = Input {
            source,
            kind: Vec::new(),
            position: Vec::new(),
        };

        let mut lex = Lexer::<TokenKind>::new(source);

        while let Some(tk) = lex.next() {
            if tk == TokenKind::CommentBlockOpen {
                let mut closed = false;
                let mut join_span = lex.span();
                while let Some(t) = lex.next() {
                    join_span.end = lex.span().end;
                    if t == TokenKind::CommentBlockClose {
                        closed = true;
                        break;
                    }
                }

                if closed {
                    input.kind.push(TokenKind::BlockComment);
                } else {
                    input.kind.push(TokenKind::Error);
                }
                input.position.push(join_span);
            } else {
                input.kind.push(tk);
                input.position.push(lex.span());
            }
        }

        input
    }

    pub fn token_value(&self, index: usize) -> Option<&'a str> {
        if index < self.kind.len() {
            Some(&self.source[self.position[index].start..self.position[index].end])
        } else {
            // return None for out of bound index
            None
        }
    }

    pub fn kind_of(&self, index: usize) -> TokenKind {
        if index < self.kind.len() {
            self.kind[index]
        } else {
            TokenKind::EOF
        }
    }

    pub fn position_of(&self, index: usize) -> Option<Range<usize>> {
        if index < self.kind.len() {
            Some(self.position[index].clone())
        } else {
            // return error for out of bound index
            None
        }
    }

    pub fn size(&self) -> usize {
        self.kind.len()
    }
}

#[cfg(test)]
mod tests {
    use super::Input;

    fn test(source: &str, snapshot_name: &str) {
        let input = Input::new(&source);

        insta::assert_yaml_snapshot!(snapshot_name, input);
    }

    #[test]
    fn test_comment_block() {
        let source = r#"
        /*a + b == 10*/
        a + 10
    "#;
        test(source, "test_comment_block");
    }

    #[test]
    fn test_comment_error() {
        let source = r#"
        pragma 2.1.1;
        /*a + b == 10*
        a + 10
        template

        /*
    "#;
        test(source, "test_comment_error");
    }

    #[test]
    fn test_pragma() {
        let source = r#"
        /* test pragma token kinds */

    pragma circom 2.0.0;

    "#;
        test(source, "test_pragma");
    }

    #[test]
    fn test_function() {
        let source = r#"
    function nbits(a) {
        var n = 1;
        var r = 0;
        while (n-1<a) {
            r++;
            n *= 2;
        }
        return r;
    }"#;
        test(source, "test_function");
    }

    #[test]
    fn test_operators() {
        let source = r#"
        ({[]})
        ;.,:
        && &
        || |
        != !
        === == =
        --> ==>
        <-- <==
        <= <
        >= >
        ++ += +
        -- -= -
        **= **
        * *=
        / /=
        \ \=
        % %=
        ^ ^=
        ~ ~=
        >> >>=
        << <<=
        & &=
        | |=
    }"#;
        test(source, "test_operators");
    }
}
