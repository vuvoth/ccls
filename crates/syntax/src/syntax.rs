use parser::grammar::entry::Scope;
use parser::input::Input;
use parser::output::{Child, Output};
use parser::parser::Parser;
use parser::token_kind::TokenKind;
use rowan::{GreenNode, GreenNodeBuilder};

pub use rowan::{
    api::Preorder, Direction, NodeOrToken, SyntaxText, TextRange, TextSize, TokenAtOffset,
    WalkEvent,
};

use crate::syntax_node::SyntaxNode;

pub struct SyntaxTreeBuilder<'a> {
    builder: GreenNodeBuilder<'static>,
    input: &'a Input<'a>,
}

impl<'a> SyntaxTreeBuilder<'a> {
    pub fn new(input: &'a Input) -> Self {
        Self {
            builder: GreenNodeBuilder::new(),
            input,
        }
    }
    pub fn build_rec(&mut self, tree: &Output) {
        self.builder.start_node(tree.kind().into());
        for child in tree.children() {
            match child {
                Child::Token(token_id) => {
                    let token_kind = self.input.kind_of(*token_id);
                    // TODO: return Error to replace .unwrap()
                    let token_value = self.input.token_value(*token_id).unwrap();
                    self.builder.start_node(token_kind.into());
                    self.builder.token(token_kind.into(), token_value);
                    self.builder.finish_node();
                }
                Child::Tree(child_tree) => self.build_rec(child_tree),
                Child::Error(error) => {
                    let token_kind = TokenKind::Error;
                    let token_value = error.as_str();

                    self.builder.start_node(token_kind.into());
                    self.builder.token(token_kind.into(), token_value);
                    self.builder.finish_node();
                }
            }
        }

        self.builder.finish_node();
    }

    pub fn build(&mut self, tree: Output) {
        self.build_rec(&tree);
    }

    pub fn finish(self) -> GreenNode {
        self.builder.finish()
    }

    pub fn syntax_tree(source: &str) -> SyntaxNode {
        let input = Input::new(source);

        let output = Parser::parsing(&input);

        let mut builder = SyntaxTreeBuilder::new(&input);
        builder.build(output);
        let green = builder.finish();
        SyntaxNode::new_root(green)
    }
}

pub fn syntax_node_from_source(source: &str, scope: Scope) -> SyntaxNode {
    let input = Input::new(&source);
    let output = Parser::parsing_with_scope(&input, scope);

    // output is a tree whose node is index of token, no content of token
    // convert output into green node
    let mut builder = SyntaxTreeBuilder::new(&input);
    builder.build(output);
    let green = builder.finish();

    // then cast green node into syntax node
    let syntax = SyntaxNode::new_root(green);

    syntax
}

#[cfg(test)]
mod tests {
    use crate::test_syntax;
    use parser::grammar::entry::Scope;

    #[test]
    fn pragma_happy_test() {
        test_syntax!("/src/test_files/happy/pragma.circom", Scope::Pragma);
    }

    #[test]
    fn template_happy_test() {
        // SOURCE & EXPECTED RESULT
        test_syntax!("/src/test_files/happy/template.circom", Scope::Template);
    }

    #[test]
    fn block_happy_test() {
        test_syntax!("/src/test_files/happy/block.circom", Scope::Block);
    }

    #[test]
    fn comment_happy_test() {
        test_syntax!(
            "/src/test_files/happy/block_comment.circom",
            Scope::CircomProgram
        );
        test_syntax!(
            "/src/test_files/happy/line_comment.circom",
            Scope::CircomProgram
        );
    }

    #[test]
    fn full_circom_program() {
        test_syntax!(
            "/src/test_files/happy/full_circom_program.circom",
            Scope::CircomProgram
        );
    }
}
