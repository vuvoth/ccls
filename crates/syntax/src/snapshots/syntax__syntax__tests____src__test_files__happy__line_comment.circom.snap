---
source: crates/syntax/src/syntax.rs
expression: "crate :: view_syntax :: view_ast(& syntax)"
---
 CircomProgram 0..45
|      EndLine 0..2
|     |      EndLine 0..2 "\r\n"
|      CommentLine 2..15
|     |      CommentLine 2..15 "// comment :>"
|      EndLine 15..17
|     |      EndLine 15..17 "\r\n"
|      EndLine 17..19
|     |      EndLine 17..19 "\r\n"
|      WhiteSpace 19..23
|     |      WhiteSpace 19..23 "    "
|      Pragma 23..43
|     |      Pragma 23..29
|     |     |      Pragma 23..29 "pragma"
|     |      WhiteSpace 29..30
|     |     |      WhiteSpace 29..30 " "
|     |      Circom 30..36
|     |     |      Circom 30..36 "circom"
|     |      WhiteSpace 36..37
|     |     |      WhiteSpace 36..37 " "
|     |      Version 37..42
|     |     |      Version 37..42 "2.0.0"
|     |      Semicolon 42..43
|     |     |      Semicolon 42..43 ";"
|      EndLine 43..45
|     |      EndLine 43..45 "\r\n"
